python/ray/serve/_private/http_proxy.py

```diff
@@ -565,113 +744,260 @@ async def __call__(self, scope, receive, send):
+        """
+        raise NotImplementedError
+
+    def setup_request_context_and_handle(
         self,
+        app_name: str,
         handle: RayServeHandle,
-        scope: Scope,
-        receive: Receive,
-        send: Send,
-    ) -> str:
-        http_body_bytes = await receive_http_body(scope, receive, send)
+        route_path: str,
+        serve_request: ServeRequest,
+    ) -> Tuple[RayServeHandle, str]:
+        """Setup the request context and handle for the request.
 
-        # NOTE(edoakes): it's important that we defer building the starlette
-        # request until it reaches the replica to avoid unnecessary
-        # serialization cost, so we use a simple dataclass here.
-        request = HTTPRequestWrapper(scope, http_body_bytes)
+        Each proxy needs to implement its own logic for setting up the request context
+        and handle.
+        """
+        raise NotImplementedError
 
-        # Perform a pickle here to improve latency. Stdlib pickle for simple
-        # dataclasses are 10-100x faster than cloudpickle.
-        request = pickle.dumps(request)
+    async def send_request_to_replica_streaming(
+        self,
+        request_id: str,
+        handle: RayServeHandle,
+        serve_request: ServeRequest,
+    ) -> ServeResponse:
+        """Send the request to the replica and handle streaming response.
 
-        retries = 0
-        backoff_time_s = 0.05
-        backoff = False
-        loop = get_or_create_event_loop()
-        # We have received all the http request conent. The next `receive`
-        # call might never arrive; if it does, it can only be `http.disconnect`.
-        while retries < HTTP_REQUEST_MAX_RETRIES + 1:
-            assignment_task: asyncio.Task = handle.remote(request)
-            client_disconnection_task = loop.create_task(receive())
-            done, _ = await asyncio.wait(
-                [assignment_task, client_disconnection_task],
-                return_when=FIRST_COMPLETED,
-            )
-            if client_disconnection_task in done:
-                message = await client_disconnection_task
-                assert message["type"] == "http.disconnect", (
-                    "Received additional request payload that's not disconnect. "
-                    "This is an invalid HTTP state."
-                )
-                logger.warning(
-                    f"Client from {scope['client']} disconnected, cancelling the "
-                    "request.",
-                    extra={"log_to_stderr": False},
-                )
-                # This will make the .result() to raise cancelled error.
-                assignment_task.cancel()
-            else:
-                client_disconnection_task.cancel()
+        Each proxy needs to implement its own logic for sending the request and
+        handling the streaming response.
+        """
+        raise NotImplementedError
 
+
+class GRPCProxy(GenericProxy):
```

Comment:
@GeneDer are you making any changes to `HTTPProxy`?

It's impossible to review right now as GH seems to assume that you're replacing HTTPProxy with GRPCProxy. 

Can you inverse the ordering of these classes so that it treats as a new code being appended instead?

Reply:
The way this works is by factoring out the original `HTTPProxy` into a `GenericProxy` that contains only things that applies to both protocols. Both `GRPCProxy` and the new `HTTPProxy` are subclass of `GenericProxy` and have their own implementation of entry points and some specific logics for dealing with their input and outputs.

Reply:
Understood. I'm referring to the name of the module -- we should rename it to just be `proxy.py` or (better) split out Generic/HTTP/gRPC proxies into their own modules

Reply:
No need to do in this PR, can do it in a follow-up